virttest/migration.py

```diff
@@ -270,27 +282,159 @@ def do_migration(self, vms, srcuri, desturi, migration_type,
+                                 "before_pause": "yes"},
+                                {"func": <function domjobabort at 0x7f5835cd08c8>,
+                                 "before_pause": "no"}
+                               ]
+            :return: boolean, True to collect events, otherwise False
+            """
+            if not funcs_to_run:
+                return False
+            if isinstance(funcs_to_run, list):
+                for one_func in funcs_to_run:
+                    if isinstance(one_func, dict):
+                        begin_event = one_func.get('begin_event')
+                        end_event = one_func.get('end_event')
+                        if any([begin_event, end_event]):
+                            return True
+                    else:
+                        # It should be a FunctionType
+                        return False
+            elif isinstance(funcs_to_run, types.FunctionType):
+                return False
+
+        def _run_funcs(funcs_to_run, before_pause, virsh_event_session=None):
+            """
+            Execute the functions during migration
+
+            :param funcs_to_run: the function or list of functions
+            :param before_pause: True to run functions before guest is
+                                 paused on source host, otherwise, False
+            :param virsh_event_session: VirshSession to collect domain events
+            :raises: exceptions.TestError if any test error happens
+            """
+            for one_func in funcs_to_run:
+                if isinstance(one_func, types.FunctionType):
+                    if not before_pause:
+                        if one_func == process.run:
+                            try:
+                                one_func(args['func_params'], shell=args['shell'])
+                            except KeyError:
+                                one_func(args['func_params'])
+                        elif one_func == virsh.migrate_postcopy:
+                            one_func(vm.name, uri=srcuri, debug=True)
+                        else:
+                            if 'func_params' in args:
+                                logging.debug("Run function {} with parameters".format(one_func))
+                                one_func(args['func_params'])
+                            else:
+                                logging.debug("Run function {}".format(one_func))
+                                one_func()
+                        return
+                    else:
+                        logging.debug("Only support to run the function "
+                                      "after guest is paused")
+                elif isinstance(one_func, dict):
+                    logging.debug("Handle function invoking:%s", one_func)
+                    before_vm_pause = 'yes' == one_func.get('before_pause', 'no')
+                    if before_vm_pause != before_pause:
```

Comment:
I think we should raise an exception if the VM is paused before executing a func that need to be run earlier...

Reply:
updated.